### PR TITLE
[MemPool 1/N] Uses MemPoolContext to route allocations from CUDACachingAllocator

### PR DESCRIPTION
**Pull request by syed-ahmed**
This PR starts plumbing in usage of `MemPoolContext::getActiveMemPool();` in CUDACachingAllocator. If there is an active mempool, we allocate using the active mempool's allocator. A context manager is also added that marks a region of pytorch code to use a certain pool.

Part of https://github.com/pytorch/pytorch/issues/124807.

### CODE REVIEW

**Comment by syed-ahmed** [comment]
@ezyang @zdevito this is ready for review. Thank you in advance!

File: docs/source/conf.py [hunk 1]
```diff
@@ -424,6 +424,7 @@
     "reset_max_memory_cached",
     "reset_peak_memory_stats",
     "set_per_process_memory_fraction",
+    "use_mem_pool",
```
Comment (albanD) [inline docs/source/conf.py:427]: Please do not add any new entry to this file and make sure to properly document this API if it is expected to be public (which I think you're already doing btw).

**Review by albanD — CHANGES_REQUESTED** [review]

File: docs/source/conf.py (hunk 1 above)
Comment (syed-ahmed) [reply docs/source/conf.py:427]: Sounds good! I'll try to figure out for real why coverage fails on my diff without an entry in this file.

Comment (syed-ahmed) [reply docs/source/conf.py:427]: Unless you can show me the right direction :)

Comment (albanD) [reply docs/source/conf.py]: Sounds good. Thanks for fixing it!

File: torch/_C/__init__.pyi.in [hunk 2]
```diff
@@ -1821,6 +1821,7 @@ def _cuda_cudaHostAllocator() -> _int: ...
 def _cuda_cudaCachingAllocator_raw_alloc(size: _int, cuda_stream: _int) -> _int: ...
 def _cuda_cudaCachingAllocator_raw_delete(ptr: _int) -> None: ...
 def _cuda_cudaCachingAllocator_set_allocator_settings(env: str) -> None: ...
+def _cuda_beginAllocateToPool(device: _int, mempool_id: Tuple[_int, _int]) -> None: ...
```
Comment (ezyang) [inline torch/_C/__init__.pyi.in:1824]: This is confusing, you still end the allocations still with `_cuda_endAllocateCurrentStreamToPool`? What exactly is the difference?

File: torch/cuda/memory.py [hunk 3]
```diff
@@ -1002,3 +1015,28 @@ def __init__(self, pool: MemPool):
...
+
+    Args:
+        pool(torch.cuda.MemPool): a MemPool object to be made active so that
+            allocations route to this pool.
+        device (torch.device or int, optional): selected device. Uses MemPool on
+            the current device, given by :func:`~torch.cuda.current_device`,
+            if :attr:`device` is ``None`` (default).
+
+    """
+    torch.cuda.init()
```
Comment (ezyang) [inline torch/cuda/memory.py:1032]: How come this is necessary?

File: c10/cuda/CUDACachingAllocator.cpp [hunk 4]
```diff
@@ -2674,7 +2674,12 @@ class DeviceCachingAllocator {
         // any potential exceptions in the cudaMallocMaybeCapturing function.
         auto sg = c10::make_scope_exit([&]() { lock.lock(); });
         lock.unlock();
-        p.err = cudaMallocMaybeCapturing(&ptr, size);
+      }
+      auto active_pool = MemPoolContext::getActiveMemPool();
+      if (active_pool && active_pool->allocator() &&
+          p.pool->owner_PrivatePool) {
```
Comment (ezyang) [inline c10/cuda/CUDACachingAllocator.cpp:2680]: For my education, why do we have to test owner_PrivatePool?

**Review by ezyang — COMMENTED** [review]
I'd like to understand the begin/end API situation better, but I don't see substantive problems here.

**Pull request closed without merging**